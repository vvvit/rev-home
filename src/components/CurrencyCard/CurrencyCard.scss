@import '../../styles/variables.scss';

.CurrencyCard {
    padding: 70px 24px;

    &_type {
        &_buy {
            background-color: $color-light-gray;
        }
    }

    &-Row {
        display: flex;

        &_justify {
            &_between {
                justify-content: space-between;
            }
        }
    }

    &-CardInfo {
        margin-top: 10px;
        font-size: $font-size-l;
    }

    &-Info_negative {
        color: $color-red;
    }
}
