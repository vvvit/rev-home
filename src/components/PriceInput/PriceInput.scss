@import '../../styles/variables.scss';

.PriceInput {
    &-Input {
        width: 100%;
        font-size: $font-size-xl;
        text-align: right;
        outline: 0;
        border: 0;
        box-sizing: border-box;
        background: transparent;
    }
}
