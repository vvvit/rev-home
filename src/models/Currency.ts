export const enum CurrencyName {
    USD = 'usd',
    EUR = 'eur',
    GBP = 'gbp',
}

export interface ICurrency {
    name: CurrencyName,
    value: number
}
