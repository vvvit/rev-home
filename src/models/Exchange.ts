export const enum CurrencyExchangeType {
    BUY = 'buy',
    SELL = 'sell'
}
