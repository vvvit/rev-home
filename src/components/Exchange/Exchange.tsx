import * as React from 'react';
import {cn} from '@bem-react/classname';

import { CurrencyCard, CurrencyCardType } from '../CurrencyCard/CurrencyCard';

const cnApp = cn('Exchange');

class Exchange extends React.Component {
  render() {
    return (
      <div className={cnApp()}>
        <CurrencyCard
            currency="RUR"
            type={CurrencyCardType.buy}
        />
        <CurrencyCard
            currency="USD"
            type={CurrencyCardType.sell}
        />
      </div>
    );
  }
}

export { Exchange };
