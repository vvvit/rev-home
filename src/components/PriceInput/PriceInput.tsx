import * as React from 'react';
import {boundMethod} from 'autobind-decorator';
import {cn} from "@bem-react/classname";

import './PriceInput.scss';

interface IPriceInputProps {
    value: string;
    onChange: (value: number) => void;
}

interface IPriceInputState {
    value: string;
}

const cnPriceInput = cn('PriceInput');

export const MAX_VALUE_LENGHT = 15;

export class PriceInput extends React.Component<IPriceInputProps, IPriceInputState> {
    static defaultProps: Partial<IPriceInputProps> = {
        onChange: () => {},
        value: '0'
    };

    state: IPriceInputState = {
        value: this.props.value
    };

    @boundMethod
    onChange(e: React.FormEvent<HTMLInputElement>) {
       e.persist();

        const value = this.formatValue(e.currentTarget.value);

        this.setState({ value });
    }

    render() {
        return (
            <div className={cnPriceInput()}>
                <input className={cnPriceInput('Input')} value={this.state.value} onChange={this.onChange} />
            </div>
        );
    }

    formatValue(value: string) {
        // valid float string without insignificant trailing zeros
        const validNumber = parseFloat(parseFloat(value).toFixed(2));
        let validString = isNaN(validNumber) ? '0' : String(validNumber);

        if (value.length > MAX_VALUE_LENGHT) {
            validString = value.substring(-value.length + MAX_VALUE_LENGHT, MAX_VALUE_LENGHT);
        } else if (validString.length < (MAX_VALUE_LENGHT - 2) && value.indexOf('.0') === validString.length) {
            // for strings like '11.0'
            validString += '.0';
        } else if (validString.length < (MAX_VALUE_LENGHT - 1) && value.indexOf('.') === validString.length) {
            // for strings like '11.'
            validString += '.';
        }

        return validString;
    }
}
