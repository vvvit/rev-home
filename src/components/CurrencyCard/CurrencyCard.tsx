import * as React from 'react';
import {boundMethod} from 'autobind-decorator';
import {cn} from '@bem-react/classname';

import {CurrencyName} from '../../models/Currency';
import {CurrencyExchangeType} from '../../models/Exchange';
import {CurrencySelect} from '../CurrencySelect/CurrencySelect';
import {exchangeConnector} from '../CurrencySelect/connectors/exchangeConnector';
import {PriceInput} from '../PriceInput/PriceInput';

import './CurrencyCard.scss';

const cnApp = cn('CurrencyCard');

const CurrencySelectContainer = exchangeConnector(CurrencySelect);

export interface ICurrencyCardProps {
    currency: CurrencyName;
    value: string;
    balance: number;
    type: CurrencyExchangeType;
    onChangeValue: (value: string, type: CurrencyExchangeType) => void;
}

export class CurrencyCard extends React.Component<ICurrencyCardProps> {
    render() {
        const {type, currency, value, balance} = this.props;

        const isNegativeBalance = type === CurrencyExchangeType.SELL && (balance < parseFloat(value));

        return (
            <div className={cnApp({type})}>
                <div className={cnApp('Row')}>
                    <CurrencySelectContainer type={type} value={currency}/>
                    <PriceInput value={value} type={type} onChange={this.onChangePrice}/>
                </div>
                <div className={cnApp('Row', {justify: 'between'}, [cnApp('CardInfo')])}>
                    <div className={cnApp('Info', {negative: isNegativeBalance})}>
                        {`Balance: ${balance} ${currency}`}
                    </div>
                </div>
            </div>
        );
    }

    @boundMethod
    onChangePrice(value: string) {
        const {onChangeValue, type} = this.props;

        onChangeValue(value, type);
    }
}
