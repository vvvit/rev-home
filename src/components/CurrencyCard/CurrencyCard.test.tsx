import * as React from 'react';
import { shallow } from 'enzyme';

import { CurrencyCard } from './CurrencyCard';
import { PriceInput } from '../PriceInput/PriceInput';
import { CurrencyExchangeType } from '../../models/Exchange';
import { CurrencyName } from '../../models/Currency';

describe('CurrencyCard', () => {
    let onChange: jest.Mock;

    beforeEach(() => {
        onChange = jest.fn();
    });

    it('renders correctly with BUY type', () => {
        expect(shallow(
            <CurrencyCard
                type={CurrencyExchangeType.BUY}
                balance={99}
                value="100.01"
                currency={CurrencyName.EUR}
                onChangeValue={onChange}
            />
        )).toMatchSnapshot();
    });

    it('renders correctly with SELL type', () => {
        expect(shallow(
            <CurrencyCard
                type={CurrencyExchangeType.SELL}
                balance={99}
                value="100.01"
                currency={CurrencyName.USD}
                onChangeValue={onChange}
            />
        )).toMatchSnapshot();
    });

    it('invokes onchange property when value changed inside input', () => {
        const component = shallow(
            <CurrencyCard
                type={CurrencyExchangeType.SELL}
                balance={99}
                value="100.01"
                currency={CurrencyName.USD}
                onChangeValue={onChange}
            />
        );

        const PriceInputComponent = component.find(PriceInput);
        // изменение цены дожно активировать услугу
        PriceInputComponent.prop('onChange')('1000');

        expect(onChange).toBeCalledWith('1000', CurrencyExchangeType.SELL);
    });
});
