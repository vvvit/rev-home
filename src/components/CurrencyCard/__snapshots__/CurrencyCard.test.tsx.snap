// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyCard renders correctly with BUY type 1`] = `
<div
  className="CurrencyCard CurrencyCard_type_buy"
>
  <div
    className="CurrencyCard-Row"
  >
    <Connect(CurrencySelect)
      type="buy"
      value="EUR"
    />
    <PriceInput
      onChange={[Function]}
      type="buy"
      value="100.01"
    />
  </div>
  <div
    className="CurrencyCard-Row CurrencyCard-Row_justify_between CurrencyCard-CardInfo"
  >
    <div
      className="CurrencyCard-Info"
    >
      Balance: 99 EUR
    </div>
  </div>
</div>
`;

exports[`CurrencyCard renders correctly with SELL type 1`] = `
<div
  className="CurrencyCard CurrencyCard_type_sell"
>
  <div
    className="CurrencyCard-Row"
  >
    <Connect(CurrencySelect)
      type="sell"
      value="USD"
    />
    <PriceInput
      onChange={[Function]}
      type="sell"
      value="100.01"
    />
  </div>
  <div
    className="CurrencyCard-Row CurrencyCard-Row_justify_between CurrencyCard-CardInfo"
  >
    <div
      className="CurrencyCard-Info CurrencyCard-Info_negative"
    >
      Balance: 99 USD
    </div>
  </div>
</div>
`;
